accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+        pattern that will match a list with a specific number of elements larger
+        than that.*
 
-3.  Otherwise, the result is a union of objects where each object includes one
-    of the changed fields and leaves the others alone. This is how we tile over
-    the holes created by subtracting `R`.
+    2.  For `i` from `0` to `n`, half-inclusive:
 
-    1.  For each field in `changed`, create an object with the same type as `L`
-        with all fields from `fixed`, this field from `changed` and all other
-        field names in `changed` set to their original space in `L`.
+        1.  A space with type `T` and an arity restriction with length `i` and
+            `hasRest` `false`.
 
-    2.  The result is the union of those spaces.
+    3.  And a space with type `T` and an arity restriction with length `n` and
+        `hasRest` `true`.
 
-    For example:
+    *For example, given:*
 
-    ```
-    (w: bool, x: bool, y: bool, z: bool) - (x: true, y: false, z: true)`
+    ```dart
+    switch (list) {
+      case [_, ..., _, _]:
+      case [_, _, _, ..., _]:
+    }
     ```
 
-    `fixed` is `(w: bool)`. `changed` is `(x: false, y: true, z: false)`. The
-    result is a union of:
+    *The expansion of the value list is:*
 
     ```
-    (x: false, y: bool, z: bool) // Set x to D.
-    (x: bool, y: true, z: bool) // Set y to D.
-    (x: bool, y: bool, z: false) // Set z to D.
+    List(length: 0, hasRest: false)
+    List(length: 1, hasRest: false)
+    List(length: 2, hasRest: false)
+    List(length: 3, hasRest: false)
+    List(length: 4, hasRest: false)
+    List(length: 4, hasRest: true)
     ```
 
-## Space intersection
+*   Otherwise, `s` does not expand and the result is just `s`.
 
-Intersection is (mercifully) simpler than subtraction. In fact, we don't even
-need to calculate the actual intersection. We just need to know if it's empty
-or not.
+The expansion procedure is applied recursively to each resulting space described
+here until no more expansions take place.
 
-To calculate if the intersection `I` of spaces `L` and `R` is empty:
+*For example, the expansion of `FutureOr<bool?>` is:*
 
-1.  If `L` or `R` is empty, then `I` is empty.
+```
+Future<bool?>
+true
+false
+null
+```
 
-2.  If `L` or `R` is a union, then `I` is empty if the intersection of every
-    arm of the union with the other operand is empty.
+### Overapproximation
 
-3.  Otherwise, we're intersecting two object spaces.
+The *overapproximation* of a type `T` is `T` with all type variables replaced
+with their *defaults*:
 
-    1.  If neither object's type is a subtype of the other then `I` is empty.
-        They are unrelated types with no (known) intersection.
+1.  If `T` is in a contravariant position, then the default is `Never`.
 
-    2.  Otherwise, go through the fields. If the intersection of any
-        corresponding pair of fields is empty, then `I` is empty.
+2.  Otherwise, the default is the bound if there is one.
```

Comment:
`class C<T, S extends T, R extends List<T>> {}` has bound declarations, but not bound types.
Use I2B, it's good for you.

Reply:
I don't understand the distinction you're making here, sorry. Can you explain, or did Johhni's explanation of how overapproximation differs from I2B address this?